packages/apollo-server-core/src/plugin/cacheControl/index.ts

```diff
@@ -34,124 +34,180 @@ declare module 'graphql/type/definition' {
-            // default maxAge, so we do it here.)
-            if (
-              (targetType instanceof GraphQLObjectType ||
-                targetType instanceof GraphQLInterfaceType ||
-                !info.path.prev) &&
-              hint.maxAge === undefined
-            ) {
-              hint.maxAge = defaultMaxAge;
-            }
 
-            const path = responsePathAsArray(info.path).join('.');
+              // If this resolver returns an object or is a root field and we haven't
+              // seen an explicit maxAge hint, set the maxAge to 0 (uncached) or the
+              // default if specified in the constructor. (Non-object fields by
+              // default are assumed to inherit their cacheability from their parents.
+              // But on the other hand, while root non-object fields can get explicit
+              // hints from their definition on the Query/Mutation object, if that
+              // doesn't exist then there's no parent field that would assign the
+              // default maxAge, so we do it here.)
+              if (
+                (targetType instanceof GraphQLObjectType ||
+                  targetType instanceof GraphQLInterfaceType ||
+                  !info.path.prev) &&
+                hint.maxAge === undefined
+              ) {
+                hint.maxAge = defaultMaxAge;
+              }
 
-            if (hint.maxAge !== undefined || hint.scope !== undefined) {
-              addHint(hints, path, hint);
-            }
+              const path = responsePathAsArray(info.path).join('.');
 
-            info.cacheControl = {
-              setCacheHint: (hint: CacheHint) => {
-                addHint(hints, path, hint);
-              },
-              cacheHint: hint,
-            };
-          },
-        }),
+              info.cacheControl = {
+                setCacheHint: (dynamicHint: CacheHint) => {
+                  hint = mergeHints(hint, dynamicHint);
+                },
+                cacheHint: hint,
+              };
 
-        // Keep this sync! See ##KeepHooksSync.
-        responseForOperation() {
-          // We are not supplying an answer, we are only setting the cache
-          // policy if it's not set! Therefore, we return null.
-          setOverallCachePolicyWhenUnset();
-          return null;
+              // When the field is done, call addHint once. By calling addHint
+              // once, we don't need to "undo" the effect on overallCachePolicy
+              // of a static hint that gets refined by a dynamic hint.
+              return () => {
+                if (hint.maxAge !== undefined || hint.scope !== undefined) {
+                  if (hints.has(path)) {
+                    throw Error(
+                      "shouldn't happen: addHint should only be called once per path",
+                    );
+                  }
+                  hints.set(path, hint);
+                  policyUpdater?.addHint(hint);
```

Comment:
With the addition of the `PolicyUpdater`, I wonder if it would be possible to get rid of the `hints` map (and the computation of the `path` string) altogether. If all we care about is the overall cache policy, couldn't we skip keeping  individual hints around?

Reply:
Yes, I think you're right. I'll give this a shot.

I think that the information contained in `cacheHints` would be valuable information to drive features we haven't built yet like partial operation caching, but it's actually quite a lot of memory overhead given that nothing uses those features. I'll remove it as its own commit on this PR which should make it feasible to revert if we want to back.

(It might be a reasonable idea to keep hints that are the same as their immediate parent out of the map, which would solve part of the overhead problem...)

Reply:
OK, I made this change. (I kept the Map with string keys but made it something that the tests ingest and that has no overhead otherwise.)